### PR TITLE
__repr__ and __str__ for models

### LINKED ISSUE #392: __str__ and __repr__ for discretizations
This would be helpful for `InputOutputSystem`s.
@sdrave Would it be useful in general for discretizations?

### PR DESCRIPTION
**Pull request by pmli**
This includes some basic `__repr__` and `__str__` methods for models. Comments are welcome.

Closes #392.

### CODE REVIEW

**Comment by sdrave** [comment]
I see that you have decided to let `__str__` and `__repr__` behave quite differently. Actually, I am not really sure what we want here. I know that it is recommended practice to let the result of `__repr__` be mostly equivalent to the Python code needed to instantiate the object while `__str__` should give a nice string representation. However, our objects might be a little too complex for such a `__repr__` implementation. In particular, if proper `__repr__` functions are implemented for operators as well. (Think of a parameterized problem, where multiple operators are (probably nested) `LincombOperators`.) Also, I am not sure if we really want to print all (optional) arguments of the construtor call.

@pymor/pymor-devs , what do you think?

**Comment by pmli** [comment]
Ok, I'm actually fine with using `__repr__` for "pretty printing" to avoid using `print`. I wasn't sure how much we want (or how much it makes sense) to follow the conventions ([`__repr__`](https://docs.python.org/3/reference/datamodel.html#object.__repr__), [`__str__`](https://docs.python.org/3/reference/datamodel.html#object.__str__)).

**Comment by sdrave** [comment]
This is really something, where I am not sure what the right thing is. Maybe someone else also has an opinion, @pymor/pymor-devs?

**Comment by renefritze** [comment]
I also think it's fine to break from the `__repr__` rule for those kinds of complex objects.

**Comment by pmli** [comment]
@pymor/pymor-devs How about first adding `__str__` in this PR and later discuss `__repr__`? I would like to hear what do you think `print(model)` should look like for `StationaryModel` and `InstationaryModel` (and `iosys` models if you have ideas).

**Comment by sdrave** [comment]
I guess for `StationaryModel` i would like something like

```python
f'''
StationaryModel {self.name if self.name != 'StationaryModel' else ''}
    parameter_space: {self.parameter_space}
    solution_space: {self.solution_space}
    operator: {self.operator}
    rhs: {self.rhs}
'''
```
If the `str` of `self.operator` or `self.rhs` has multiple lines, it should be indented accordingly.

File: src/pymor/models/basic.py [hunk 1]
```diff
@@ -29,6 +28,13 @@ def __init__(self, products=None, estimator=None, visualizer=None,
                 setattr(self, f'{k}_product', v)
                 setattr(self, f'{k}_norm', induced_norm(v))
 
+    def __repr__(self):
+        strings = [f'{self.__class__.__name__}(\n']
+        for parameter in self.__signature__.parameters:
+            strings.append(f'    {parameter}={getattr(self, parameter)!r},\n')
+        strings.append(')')
+        return ''.join(strings)
```
Comment (sdrave) [inline src/pymor/models/basic.py]: If we want to have such a method, why only for `Models`? We might consider moving it to `BasicInterface`. Not sure what the consequences would be, however.

File: src/pymor/models/basic.py [hunk 2]
```diff
@@ -122,13 +128,43 @@ def __init__(self, operator, rhs, outputs=None, products=None,
         super().__init__(products=products,
                          estimator=estimator, visualizer=visualizer,
                          cache_region=cache_region, name=name)
+        self.linear = operator.linear
```
Comment (sdrave) [inline src/pymor/models/basic.py]: Maybe this should be moved to a different PR? There should also be a check that `rhs` is actually linear.

File: src/pymor/models/basic.py [hunk 3]
```diff
@@ -242,6 +278,7 @@ def __init__(self, T, initial_data, operator, rhs, mass=None, time_stepper=None,
                          visualizer=visualizer, cache_region=cache_region, name=name)
         self.T = T
         self.initial_data = initial_data
+        self.linear = operator.linear
```
Comment (sdrave) [inline src/pymor/models/basic.py]: See above. Also we should probably check for linearity of `mass` and `initial_data`.

File: src/pymor/models/basic.py [hunk 4]
```diff
@@ -252,6 +289,16 @@ def __init__(self, T, initial_data, operator, rhs, mass=None, time_stepper=None,
...
 
+    def __str__(self):
+        return (
+            f'{self.name}\n'
+            f'    class: {self.__class__.__name__}\n'
+            f'    parameter type: {self.parameter_type}\n'
+            f'    {"" if self.linear else "non-"}linear\n'
+            f'    number of equations: {self.solution_space.dim}\n'
+            f'    number of outputs:   {len(self.outputs)}'
+        )
```
Comment (sdrave) [inline src/pymor/models/basic.py]: This is quite different from `StationaryModel.__str__`. I would except both to return a very similar representation.

File: src/pymor/models/iosys.py [hunk 5]
```diff
@@ -800,13 +812,24 @@ class TransferFunction(InputOutputModel):
         The complex derivative of the transfer function.
     """
 
-    def __init__(self, input_space, output_space, H, dH, cont_time=True, cache_region='memory', name=None):
+    def __init__(self, input_space, output_space, tf, dtf, cont_time=True, cache_region='memory', name=None):
```
Comment (sdrave) [inline src/pymor/models/iosys.py]: This is an interface change, so a separate PR would make the change more visible.

File: src/pymor/models/basic.py (hunk 2 above)
Comment (pmli) [reply src/pymor/models/basic.py]: This was used for the previous version of `StationaryModel.__str__`, which also says if the problem is linear or not.

File: src/pymor/models/basic.py (hunk 4 above)
Comment (pmli) [reply src/pymor/models/basic.py]: You only said how you would expect `StationaryModel.__str__` to look like...

File: src/pymor/models/iosys.py (hunk 5 above)
Comment (pmli) [reply src/pymor/models/iosys.py]: This is only to make the `__repr__` work, which uses `__signature__`. Actually, another/better option would be to make `H` and `dH` the attributes (instead of `tf` and `dtf`)...

File: src/pymor/models/basic.py (hunk 4 above)
Comment (sdrave) [reply src/pymor/models/basic.py]: Well, I guess I didn't have the bigger picture in mind when saying this ..

File: src/pymor/models/iosys.py (hunk 5 above)
Comment (sdrave) [reply src/pymor/models/iosys.py]: Still, it is an interface change which should not be buried in a PR with this title.

File: src/pymor/models/basic.py (hunk 2 above)
Comment (pmli) [reply src/pymor/models/basic.py]: What I meant to ask is, do we need the `linear` attribute?

File: src/pymor/models/basic.py (hunk 4 above)
Comment (pmli) [reply src/pymor/models/basic.py]: That would be best. I could then adjust the `iosys` models.

**Comment by pmli** [comment]
How about separating this PR into one about `__repr__` and one about `__str__`?

**Comment by sdrave** [comment]
With #706 merged, how do we want to proceed here? I think it is clear that we won't need a custom `__repr__` for `Models`. What about `__str__`? I guess I would be fine with just using the `__repr__` for now. Does anybody think we need something prettier? If yes, do we want recursion in `__str__` similar to what is done in `__repr__`? (I would say that if we want a custom `__str__`, then we should actually avoid recursion and go for a short representation.)

**Comment by pmli** [comment]
I would like to have a prettier `__str__` for system classes, where it's easier to see the number of inputs, output, and states (and some other useful information).

**Comment by sdrave** [comment]
Ok, then I would propose to change `__str__` of `StationaryModel` and `InstationaryModel` to not print the contained operators.

Can you rebase on master? Then I would make the changes to these methods.

**Comment by sdrave** [comment]
@pmli, made the changes. Please have a look.

**Comment by pmli** [comment]
@sdrave I've copied the `parameter_space` and `solution_space` parts to `iosys` models and did some reordering. Let me know what you think.

**Comment by sdrave** [comment]
LGTM.
